src/providers/idp/idp_auth.c

```diff
@@ -0,0 +1,543 @@
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    extra_args[c] = talloc_asprintf(extra_args,
+                                    "--scope=%s",
+                                    idp_auth_ctx->scope);
+    if (extra_args[c] == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    chain_id = sss_chain_id_get();
+    extra_args[c] = talloc_asprintf(extra_args,
+                                    "--chain-id=%lu",
+                                    chain_id);
+    if (extra_args[c] == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    if (DEBUG_IS_SET(SSSDBG_TRACE_LIBS)) {
+        extra_args[c] = talloc_asprintf(extra_args, "--libcurl-debug");
+        if (extra_args[c] == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+        c++;
+    }
+
+    extra_args[c] = NULL;
+
+    *oidc_child_extra_args = extra_args;
+
+    ret = EOK;
+
+done:
+
+    if (ret != EOK) {
+        talloc_free(extra_args);
+    }
+
+    return ret;
+}
+
+static const char *get_stored_request_data(TALLOC_CTX *mem_ctx,
+                                           struct idp_auth_ctx *idp_auth_ctx,
+                                           struct pam_data *pd)
+{
+    int ret;
+    const char *send_data;
+    struct idp_open_req_data *open_req = NULL;
+    const char *user_code;
+    size_t user_code_len;
+
+    if (sss_authtok_get_type(pd->authtok) != SSS_AUTHTOK_TYPE_OAUTH2) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unexpected authentication type [%d][%s].\n",
+              sss_authtok_get_type(pd->authtok),
+              sss_authtok_type_to_str(sss_authtok_get_type(pd->authtok)));
+        goto done;
+    }
+
+    ret = sss_authtok_get_oauth2(pd->authtok, &user_code, &user_code_len);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to extract user code.\n");
+        goto done;
+    }
+
+    open_req = sss_ptr_hash_lookup(idp_auth_ctx->open_request_table,
+                                   user_code, struct idp_open_req_data);
+    if (open_req == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to retrieve stored request data.\n");
+        goto done;
+    }
+    if (open_req->device_code_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Missing device code data.\n");
+        goto done;
+    }
+
+    send_data = talloc_asprintf(mem_ctx, "%s\n%s",
+                                dp_opt_get_cstring(idp_auth_ctx->idp_options,
+                                                   IDP_CLIENT_SECRET),
+                                open_req->device_code_data);
+    if (send_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to generate auth data.\n");
+        goto done;
+    }
+
+done:
+    talloc_free(open_req);
```

Comment:
Is the intention to delete entry from the hash table?

Reply:
How will `talloc_free(open_req);` remove entry from the table?

'sss_ptr_hash.c' code isn't easy reading, but IIUC payload doesn't have a d-tor, only (internal) wrapper 'sss_ptr_hash_value' does:
https://github.com/SSSD/sssd/blob/3d278ec5d6a41e51a9d31d646cc165a25d63f108/src/util/sss_ptr_hash.c#L107

Imo, `sss_ptr_hash_delete()` should be used explicitly.